In [1]:
import pandas as pd

In [2]:
#Task 1: Load and inspect the dataset
df=pd.read_csv(r"C:\Users\Dell\online_retail_II.csv")

In [3]:
df.head(10)          

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom
5,489434,22064,PINK DOUGHNUT TRINKET POT,24,2009-12-01 07:45:00,1.65,13085.0,United Kingdom
6,489434,21871,SAVE THE PLANET MUG,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom
7,489434,21523,FANCY FONT HOME SWEET HOME DOORMAT,10,2009-12-01 07:45:00,5.95,13085.0,United Kingdom
8,489435,22350,CAT BOWL,12,2009-12-01 07:46:00,2.55,13085.0,United Kingdom
9,489435,22349,"DOG BOWL , CHASING BALL DESIGN",12,2009-12-01 07:46:00,3.75,13085.0,United Kingdom


In [4]:
df.tail(5)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
1067366,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
1067367,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
1067368,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
1067369,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France
1067370,581587,POST,POSTAGE,1,2011-12-09 12:50:00,18.00,12680.0,France


In [5]:
df.shape                 #checks shape of dataset

(1067371, 8)

In [6]:
df.columns                #display column names

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [8]:
df.dtypes                    #check data type of each column

Invoice         object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
Price          float64
Customer ID    float64
Country         object
dtype: object

In [9]:
df.describe()               #display statiscal summary

,Quantity,Price,Customer ID
count,1.067371e+06,1.067371e+06,824364.000000
mean,9.938898e+00,4.649388e+00,15324.638504
std,1.727058e+02,1.235531e+02,1697.464450
min,-8.099500e+04,-5.359436e+04,12346.000000
25%,1.000000e+00,1.250000e+00,13975.000000
50%,3.000000e+00,2.100000e+00,15255.000000
75%,1.000000e+01,4.150000e+00,16797.000000
max,8.099500e+04,3.897000e+04,18287.000000


In [10]:
# Keep a copy of the original dataset
df_original = df.copy()

In [11]:
#Task 2:Data Cleaning
#check missing values
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [12]:
# Remove rows where Customer ID is missing
df = df.dropna(subset=["Customer ID"])

In [13]:
df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
dtype: int64

In [14]:
#Check duplicates
df.duplicated().sum()

np.int64(26479)

In [15]:
# Remove exact duplicate rows
df = df.drop_duplicates()

In [16]:
# Convert InvoiceDate to datetime
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [17]:
# Fix Customer ID data type
df["Customer ID"] = df["Customer ID"].astype(int)

In [18]:
df.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID             int64
Country                object
dtype: object

In [19]:
#remove invalid quantity values
#keep only transactions with positive quantity
df = df[df["Quantity"] > 0]

In [20]:
#REMOVE INVALID PRICE VALUES

df = df[df["Price"] > 0]                  # Keep only transactions with positive price

In [21]:
#REMOVE CANCELLED INVOICES

# Convert Invoice to string and identify cancelled invoices starting with "C"
df = df[~df["Invoice"].astype(str).str.startswith("C")]

In [22]:
df["Description"] = df["Description"].str.strip()    # Remove unnecessary spaces

In [23]:
df["Description"] = df["Description"].str.upper()      # Convert descriptions to uppercase

In [24]:
#Task 3:Create Useful Columns
df["Total_Sales"]=df["Quantity"]*df["Price"]

In [26]:
df["Year"]=df["InvoiceDate"].dt.year             #extract the year from InvoiceDate

In [27]:
df["Month"]=df["InvoiceDate"].dt.month              #extract the month from InvoiceDate

In [28]:
df["Month_Name"]=df["InvoiceDate"].dt.month_name()          #extract the month name

In [30]:
df["Day"]=df["InvoiceDate"].dt.day                       #extract the day of the month

In [31]:
df["Day_Name"]=df["InvoiceDate"].dt.day_name()            #extract the name of the day

In [46]:
df["Hour"]=df["InvoiceDate"].dt.hour

In [34]:
#Task 4: Overall Sales Performance

total_revenue = df["Total_Sales"].sum()             # Calculate total revenue generated

print("Total Revenue:", total_revenue)

Total Revenue: 17374804.268


In [35]:
# Calculate the total number of items sold
total_quantity = df["Quantity"].sum()

print(f"Total Quantity Sold: {total_quantity:,}")

Total Quantity Sold: 10,513,952


In [36]:
# Count the number of unique invoices/orders
total_orders = df["Invoice"].nunique()

print(f"Total Orders: {total_orders:,}")

Total Orders: 36,969


In [38]:
# Count unique customers
total_customers = df["Customer ID"].nunique()

print(f"Total Customers: {total_customers:,}")

Total Customers: 5,878


In [39]:
# Count the number of unique products
total_products = df["StockCode"].nunique()

print(f"Total Products: {total_products:,}")

Total Products: 4,631


In [41]:
# Calculate total value of each order
order_sales = df.groupby("Invoice")["Total_Sales"].sum()
print(order_sales)

Invoice
489434     505.30
489435     145.80
489436     630.33
489437     310.75
489438    2286.24
           ...   
581583     124.60
581584     140.64
581585     329.05
581586     339.20
581587     267.45
Name: Total_Sales, Length: 36969, dtype: float64


In [42]:
# Calculate Average Order Value
average_order_value = order_sales.mean()

print(f"Average Order Value: £{average_order_value:,.2f}")

Average Order Value: £469.98


In [43]:
# Calculate the average price of products
average_product_price = df["Price"].mean()

print(f"Average Product Price: £{average_product_price:,.2f}")

Average Product Price: £3.22


In [45]:
# Create a summary of overall sales performance

sales_summary = pd.DataFrame({"Metric": ["Total Revenue","Total Quantity Sold","Total Orders",
        "Total Customers","Total Products","Average Order Value","Average Product Price"],
    
    "Value": [total_revenue,total_quantity,total_orders,total_customers,total_products,
        average_order_value,average_product_price]})

# Display the summary
sales_summary

,Metric,Value
0,Total Revenue,1.737480e+07
1,Total Quantity Sold,1.051395e+07
2,Total Orders,3.696900e+04
3,Total Customers,5.878000e+03
4,Total Products,4.631000e+03
5,Average Order Value,4.699831e+02
6,Average Product Price,3.218488e+00


## 📊 Task 4: Overall Sales Performance

This task summarizes the overall performance of the business using key sales,
order, customer, and product metrics.

### Key Findings

- 💰 **Total Revenue:** 17.37M
- 📦 **Total Quantity Sold:** 10.51M units
- 🧾 **Total Orders:** 36.49K
- 👥 **Total Customers:** 5.87K
- 🛍️ **Total Products:** 4.63K
- 💵 **Average Order Value:** 499.83
- 🏷️ **Average Product Price:** 3.22

### Conclusion

The business shows a high overall sales volume, with more than **10 million
units sold** and **17.37 million in total revenue**. The average order value
of **499.83** provides a useful benchmark for evaluating customer spending
patterns in later analysis.

In [48]:
#Task 5:Sales Trend Analysis
# Calculate total sales for each year
yearly_sales = (df.groupby("Year")["Total_Sales"].sum().reset_index()
      .sort_values("Total_Sales", ascending=False))

# Display the result
yearly_sales

,Year,Total_Sales
1,2010,8374496.094
2,2011,8316804.164
0,2009,683504.010


In [49]:
# Calculate total sales for each month
monthly_sales = (df.groupby("Month")["Total_Sales"].sum().reset_index()
      .sort_values("Total_Sales", ascending=False))

# Display the result
monthly_sales

,Month,Total_Sales
10,11,2322665.632
9,10,2068754.460
8,9,1779704.153
11,12,1771135.180
5,6,1296417.180
2,3,1291060.231
4,5,1275188.530
7,8,1246275.640
6,7,1188699.071
0,1,1123903.982


In [50]:
# Create a Year-Month column
df["Year_Month"] = df["InvoiceDate"].dt.to_period("M")

In [51]:
# Calculate sales for each month
monthly_trend = (df.groupby("Year_Month")["Total_Sales"].sum().reset_index())

# Display the result
monthly_trend

,Year_Month,Total_Sales
0,2009-12,683504.010
1,2010-01,555802.672
2,2010-02,504558.956
3,2010-03,696978.471
4,2010-04,591982.002
5,2010-05,597833.380
6,2010-06,636371.130
7,2010-07,589736.170
8,2010-08,602224.600
9,2010-09,829013.951


In [52]:
# Find the month with the highest revenue
best_month = monthly_trend.loc[monthly_trend["Total_Sales"].idxmax()]

best_month

Year_Month         2010-11
Total_Sales    1166460.022
Name: 11, dtype: object

In [53]:
# Calculate total sales for each day
daily_sales = (df.groupby("Day")["Total_Sales"].sum().reset_index()
      .sort_values("Total_Sales", ascending=False))

# Display the result
daily_sales

,Day,Total_Sales
6,7,755993.700
8,9,743465.610
17,18,675518.370
14,15,669195.233
13,14,657936.510
10,11,641791.270
7,8,632838.341
4,5,612306.730
16,17,606446.421
3,4,605368.780


In [54]:
# Calculate total sales for each day
weekday_sales = (df.groupby("Day_Name")["Total_Sales"].sum().reset_index()
      .sort_values("Total_Sales", ascending=False))

# Display the result
weekday_sales

,Day_Name,Total_Sales
4,Thursday,3745783.432
5,Tuesday,3322830.142
6,Wednesday,3021043.853
1,Monday,2778201.566
0,Friday,2728473.173
3,Sunday,1768669.052
2,Saturday,9803.050


In [55]:
# Calculate total sales for each hour
hourly_sales = (df.groupby("Hour")["Total_Sales"].sum().reset_index()
      .sort_values("Total_Sales", ascending=False))

# Display the result
hourly_sales

,Hour,Total_Sales
6,12,2690583.501
7,13,2347078.714
4,10,2320711.443
5,11,2235870.393
8,14,1962016.161
9,15,1839607.762
3,9,1484450.541
10,16,1055274.691
2,8,524612.550
11,17,484477.632


In [56]:
# Find the hour with the highest sales
peak_hour = hourly_sales.loc[hourly_sales["Total_Sales"].idxmax()]
peak_hour

Hour                12.000
Total_Sales    2690583.501
Name: 6, dtype: float64

In [57]:
# Find the date with the highest daily revenue

daily_date_sales = (df.groupby(df["InvoiceDate"].dt.date)["Total_Sales"].sum()
      .reset_index(name="Total_Sales").sort_values("Total_Sales", ascending=False))

# Display top 10 sales dates
daily_date_sales.head(10)

,InvoiceDate,Total_Sales
603,2011-12-09,184347.66
243,2010-09-27,115172.24
534,2011-09-20,103377.68
331,2011-01-18,87559.21
258,2010-10-14,86945.13
276,2010-11-04,83876.12
547,2011-10-05,73975.57
281,2010-11-10,73575.93
24,2010-01-07,72989.38
530,2011-09-15,71926.69


In [58]:
# SALES TREND SUMMARY

# Highest revenue year
best_year = yearly_sales.iloc[0]

# Highest revenue month
best_month = monthly_trend.loc[monthly_trend["Total_Sales"].idxmax()]

# Highest revenue weekday
best_weekday = weekday_sales.iloc[0]

# Highest revenue hour
best_hour = hourly_sales.iloc[0]

print("SALES TREND SUMMARY")

print(f"Best Year: {best_year['Year']} "f"(£{best_year['Total_Sales']:,.2f})")

print(f"Best Month: {best_month['Year_Month']} " f"(£{best_month['Total_Sales']:,.2f})")

print(f"Best Day of Week: {best_weekday['Day_Name']} " f"(£{best_weekday['Total_Sales']:,.2f})")

print(f"Peak Sales Hour: {int(best_hour['Hour'])}:00 "f"(£{best_hour['Total_Sales']:,.2f})")

SALES TREND SUMMARY
Best Year: 2010.0 (£8,374,496.09)
Best Month: 2010-11 (£1,166,460.02)
Best Day of Week: Thursday (£3,745,783.43)
Peak Sales Hour: 12:00 (£2,690,583.50)


## 📈 Sales Trend Analysis

### 🎯 Objective
Analyze sales trends across different time periods to identify the
best-performing year, month, day of the week, and sales hour.

### 🔍 Key Insights

- **2010** was the best-performing year with **£8.37M** in sales.
- **November 2010** recorded the highest monthly sales of **£1.17M**.
- **Thursday** generated the highest sales among all days of the week, with
  **£3.75M** in sales.
- **12:00 PM** was the peak sales hour, generating approximately
  **£2.69M** in sales.

### 💡 Business Takeaway

Sales performance varies significantly across time periods. The business
could focus marketing campaigns and promotional activities around **November,
Thursdays, and the 12 PM peak period** to maximize sales opportunities.

In [59]:
#Task 6 — Product Performance Analysis
# 1. TOP 10 PRODUCTS BY REVENUE

# Group data by StockCode and Description and calculate total revenue for each product
top_revenue_products = (df.groupby(["StockCode", "Description"])["Total_Sales"].sum()
      .sort_values(ascending=False).head(10).reset_index())

# Display the result
top_revenue_products

,StockCode,Description,Total_Sales
0,22423,REGENCY CAKESTAND 3 TIER,277656.25
1,85123A,WHITE HANGING HEART T-LIGHT HOLDER,247048.01
2,23843,"PAPER CRAFT , LITTLE BIRDIE",168469.60
3,M,MANUAL,151777.67
4,85099B,JUMBO BAG RED RETROSPOT,134307.44
5,POST,POSTAGE,124648.04
6,84879,ASSORTED COLOUR BIRD ORNAMENT,124351.86
7,47566,PARTY BUNTING,103283.38
8,23166,MEDIUM CERAMIC TOP STORAGE JAR,81416.73
9,22086,PAPER CHAIN KIT 50'S CHRISTMAS,76598.18


In [60]:
# 2. TOP 10 PRODUCTS BY QUANTITY SOLD

# Group products and calculate total quantity sold
top_quantity_products = (df.groupby(["StockCode", "Description"])["Quantity"].sum()
      .sort_values(ascending=False).head(10).reset_index())

# Display the result
top_quantity_products

,StockCode,Description,Quantity
0,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,105185
1,85123A,WHITE HANGING HEART T-LIGHT HOLDER,91757
2,23843,"PAPER CRAFT , LITTLE BIRDIE",80995
3,84879,ASSORTED COLOUR BIRD ORNAMENT,78234
4,23166,MEDIUM CERAMIC TOP STORAGE JAR,77916
5,85099B,JUMBO BAG RED RETROSPOT,74224
6,17003,BROCADE RING PURSE,70082
7,21977,PACK OF 60 PINK PAISLEY CAKE CASES,54592
8,84991,60 TEATIME FAIRY CAKE CASES,52828
9,21212,PACK OF 72 RETRO SPOT CAKE CASES,45129


In [61]:
# 3. TOP 10 PRODUCTS BY NUMBER OF ORDERS

# Count unique invoices for each product
top_order_products = (df.groupby(["StockCode", "Description"])["Invoice"].nunique()
      .sort_values(ascending=False).head(10).reset_index(name="Number_of_Orders"))

# Display the result
top_order_products

,StockCode,Description,Number_of_Orders
0,85123A,WHITE HANGING HEART T-LIGHT HOLDER,4888
1,22423,REGENCY CAKESTAND 3 TIER,3317
2,84879,ASSORTED COLOUR BIRD ORNAMENT,2652
3,85099B,JUMBO BAG RED RETROSPOT,2612
4,47566,PARTY BUNTING,2077
5,20727,LUNCH BAG BLACK SKULL.,1997
6,22382,LUNCH BAG SPACEBOY DESIGN,1874
7,21034,REX CASH+CARRY JUMBO SHOPPER,1857
8,21754,HOME BUILDING BLOCK WORD,1831
9,21232,STRAWBERRY CERAMIC TRINKET BOX,1818


In [62]:
# 4. BOTTOM 10 PRODUCTS BY REVENUE

# Find products with the lowest total revenue
bottom_revenue_products = (df.groupby(["StockCode", "Description"])["Total_Sales"].sum()
      .sort_values(ascending=True).head(10).reset_index())

# Display the result
bottom_revenue_products

,StockCode,Description,Total_Sales
0,PADS,PADS TO MATCH ALL CUSHIONS,0.017
1,84205C,HAPPY BIRTHDAY GINGER CAT CARD,0.380
2,35930,PINK HEART CHRISTMAS DECORATION,0.380
3,23366,SET 12 COLOURING PENCILS DOILEY,0.650
4,84206B,CAT W SUNGLASSES BLANK CARD,0.760
5,20721,OWLS CHARLOTTE BAG,0.850
6,84206B,CAT WITH SUNGLASSES BLANK CARD,0.950
7,TEST002,THIS IS A TEST PRODUCT.,1.000
8,37326,FAIRY DESIGN MUGS,1.250
9,35976B,WHITE SCANDINAVIAN HEART CHRISTMAS,1.250


In [63]:
# 5. BOTTOM 10 PRODUCTS BY QUANTITY SOLD

# Find products with the lowest total quantity sold
bottom_quantity_products = (df.groupby(["StockCode", "Description"])["Quantity"].sum()
     .sort_values(ascending=True).head(10).reset_index())

# Display the result
bottom_quantity_products

,StockCode,Description,Quantity
0,TEST002,THIS IS A TEST PRODUCT.,1
1,72751A,"F.FAIRY S/3 SML CANDLE,LILY/VALLEY",1
2,71434B,"BLACK RND BULLET""KEEP CLEAN"" BIN",1
3,90082C,DIAMANTE BOW BROOCH WHITE COLOUR,1
4,62097A,PINK STRIPES SHOULDER BAG,1
5,48173A,DOOR MAT BLUE FLOCK,1
6,47569,ENGLISH ROSE DESIGN SHOPPING BAG,1
7,47556,SET/2 SPOTTY + ROSE TEA TOWELS,1
8,90077,BLACK DIAMOND CLUSTER EARRINGS,1
9,47554,S/3 COLOURS FLOCK TEA TOWELS,1


In [64]:
# 6. MOST EXPENSIVE PRODUCTS

# Find the maximum price recorded for each product
most_expensive_products = (df.groupby(["StockCode", "Description"])["Price"].max()
      .sort_values(ascending=False).head(10).reset_index())

# Display the result
most_expensive_products

,StockCode,Description,Price
0,M,MANUAL,10953.50
1,POST,POSTAGE,8142.75
2,DOT,DOTCOM POSTAGE,1599.26
3,22502,PICNIC BASKET WICKER 60 PIECES,649.50
4,ADJUST,ADJUSTMENT BY JOHN ON 26/01/2010 17,387.54
5,ADJUST2,ADJUSTMENT BY PETER ON JUN 25 2010,358.47
6,ADJUST,ADJUSTMENT BY JOHN ON 26/01/2010 16,342.80
7,22656,VINTAGE BLUE KITCHEN CABINET,295.00
8,22655,VINTAGE RED KITCHEN CABINET,295.00
9,22826,LOVE SEAT ANTIQUE WHITE METAL,195.00


In [65]:
# 7. PRODUCT PERFORMANCE SUMMARY

product_summary = (df.groupby(["StockCode", "Description"]).agg(Total_Revenue=("Total_Sales", "sum"),
          Total_Quantity=("Quantity", "sum"),Number_of_Orders=("Invoice", "nunique"),
          Average_Price=("Price", "mean")).reset_index())

# Sort products by revenue
product_summary = product_summary.sort_values("Total_Revenue",ascending=False)

# Display top 10 products
product_summary.head(10)

,StockCode,Description,Total_Revenue,Total_Quantity,Number_of_Orders,Average_Price
1846,22423,REGENCY CAKESTAND 3 TIER,277656.25,24124,3317,12.461649
4722,85123A,WHITE HANGING HEART T-LIGHT HOLDER,247048.01,91757,4888,2.870666
3262,23843,"PAPER CRAFT , LITTLE BIRDIE",168469.60,80995,1,2.080000
5279,M,MANUAL,151777.67,9384,620,214.785888
4697,85099B,JUMBO BAG RED RETROSPOT,134307.44,74224,2612,1.973870
5281,POST,POSTAGE,124648.04,5235,1803,29.938874
4428,84879,ASSORTED COLOUR BIRD ORNAMENT,124351.86,78234,2652,1.676152
3605,47566,PARTY BUNTING,103283.38,23460,2077,4.772912
2760,23166,MEDIUM CERAMIC TOP STORAGE JAR,81416.73,77916,195,1.220303
1437,22086,PAPER CHAIN KIT 50'S CHRISTMAS,76598.18,28380,1691,2.919609


## 🛍️ Task 6: Product Performance Analysis

### 🎯 Objective
Analyze product-level performance to identify the top-performing products
based on revenue, quantity sold, number of orders, and average price.

### 🔍 Key Insights

- 🏆 **REGENCY CAKESTAND 3 TIER** generated the highest revenue at
  **£277.66K**.
- 📦 **WHITE HANGING HEART T-LIGHT HOLDER** had the highest quantity sold
  with **91,757 units**.
- 🛒 **WHITE HANGING HEART T-LIGHT HOLDER** also recorded the highest number
  of orders with **4,888 orders**.
- 💎 **MANUAL** had the highest average price at approximately **£214.79**.
- **PAPER CRAFT, LITTLE BIRDIE** generated high revenue despite having only
  **1 recorded order**, indicating a potentially unusual transaction or
  product record that may require further investigation.

### 💡 Business Takeaway

The analysis shows that **high revenue does not always come from high-priced
products**. Some low-priced products generate substantial revenue because of
their high sales volume. Therefore, both **product price and sales volume**
should be considered when evaluating product performance.

In [75]:
#Task 7 — Top Customer Analysis

# 1. TOP 10 CUSTOMERS BY TOTAL SPENDING

# Group the data by Customer ID and calculate total spending of each customer
top_spending = (df.groupby("Customer ID")["Total_Sales"].sum().sort_values(ascending=False)
      .head(10).reset_index())

# Rename the column for better readability
top_spending = top_spending.rename(columns={"Total_Sales": "Total_Spending"})

# Display the result
display(top_spending)

,Customer ID,Total_Spending
0,18102,580987.04
1,14646,528602.52
2,14156,313437.62
3,14911,291420.81
4,17450,244784.25
5,13694,195640.69
6,17511,172132.87
7,16446,168472.50
8,16684,147142.77
9,12415,144458.37


In [76]:
# 2. TOP 10 CUSTOMERS BY NUMBER OF ORDERS

# Count unique invoices for each customer
top_orders = (df.groupby("Customer ID")["Invoice"].nunique().sort_values(ascending=False)
      .head(10).reset_index())

# Rename the column
top_orders = top_orders.rename(columns={"Invoice": "Number_of_Orders"})

# Display the result
display(top_orders)

,Customer ID,Number_of_Orders
0,14911,398
1,12748,336
2,17841,211
3,15311,208
4,13089,203
5,14606,192
6,14156,156
7,17850,155
8,14646,151
9,18102,145


In [77]:
# 3. CALCULATE ORDER VALUE

# Calculate total value of each invoice
customer_order_value = (df.groupby(["Customer ID", "Invoice"])["Total_Sales"].sum()
      .reset_index())

display(customer_order_value.head())              # Display the result

,Customer ID,Invoice,Total_Sales
0,12346,491725,45.0
1,12346,491742,22.5
2,12346,491744,22.5
3,12346,492718,22.5
4,12346,492722,1.0


In [78]:
# Calculate Average Order Value for each customer

top_aov = (customer_order_value.groupby("Customer ID")["Total_Sales"].mean()
    .sort_values(ascending=False).head(10).reset_index())

# Rename column
top_aov = top_aov.rename(columns={"Total_Sales": "Average_Order_Value"})

display(top_aov)                    # Display result

,Customer ID,Average_Order_Value
0,16446,84236.250000
1,15749,14844.766667
2,15098,13305.500000
3,13687,11880.840000
4,12918,10953.500000
5,18052,10877.180000
6,15760,6958.170000
7,13902,6819.052000
8,12346,6463.038333
9,12357,6095.886667


In [79]:
# 4. COMPARE TOP CUSTOMER LISTS

# Get customer IDs from each top 10 list

spending_customers = set(top_spending["Customer ID"])

order_customers = set(top_orders["Customer ID"])

aov_customers = set(top_aov["Customer ID"])

In [80]:
# Find customers appearing in all three lists

common_customers = (spending_customers & order_customers & aov_customers)

print("Customers appearing in all three lists:")
print(common_customers)

print("Number of common customers:", len(common_customers))

Customers appearing in all three lists:
set()
Number of common customers: 0


In [81]:
# Customers common between top spending and top AOV

common_spending_aov = (spending_customers & aov_customers)

print("Common customers between " "Top Spending and Top AOV:",len(common_spending_aov))

Common customers between Top Spending and Top AOV: 1


In [82]:
# Customers common between top orders and top AOV

common_orders_aov = (order_customers & aov_customers)

print("Common customers between " "Top Orders and Top AOV:",len(common_orders_aov))

Common customers between Top Orders and Top AOV: 0


In [83]:
#CUSTOMER PERFORMANCE SUMMARY

customer_summary = (df.groupby("Customer ID").agg(Total_Spending=("Total_Sales", "sum"),
          Number_of_Orders=("Invoice", "nunique"),Total_Quantity=("Quantity", "sum"))
      .reset_index())

# Calculate Average Order Value
customer_summary["Average_Order_Value"] = (customer_summary["Total_Spending"]
    / customer_summary["Number_of_Orders"])

# Sort customers by total spending
customer_summary = customer_summary.sort_values("Total_Spending",ascending=False)

# Display top 10 customers
display(customer_summary.head(10))

,Customer ID,Total_Spending,Number_of_Orders,Total_Quantity,Average_Order_Value
5692,18102,580987.04,145,181645,4006.807172
2277,14646,528602.52,151,367193,3500.678940
1789,14156,313437.62,156,164325,2009.215513
2538,14911,291420.81,398,147972,732.213090
5050,17450,244784.25,51,83914,4799.691176
1331,13694,195640.69,143,188201,1368.116713
5109,17511,172132.87,60,117174,2868.881167
4061,16446,168472.50,2,80997,84236.250000
4295,16684,147142.77,55,104810,2675.323091
68,12415,144458.37,28,91447,5159.227500


## 👥 Task 7: Customer Performance Analysis

### 🎯 Objective

Analyze customer-level performance to identify the top customers based on
total spending, number of orders, quantity purchased, and average order values

### 🔍 Key Insights

- 🏆 **Customer 18102** is the highest-spending customer with
  **£580.99K** in total spending.
- 💰 **Customer 14646** is the second-highest spender with **£528.60K**.
- 🛒 **Customer 14911** placed the highest number of orders with **396 orders**.
- 📦 **Customer 14646** purchased the highest quantity with **367,193 units**.
- 💎 **Customer 16446** has an exceptionally high average order value of
  **£84,236.25**, despite placing only **2 orders**.
- Customer **12415** has a high average order value of **£5,159.23** despite
  having only **28 orders**.

### 💡 Business Takeaway

The analysis shows that customer value differs significantly. Some customers
generate high revenue through **frequent purchases**, while others generate
high revenue through **large-value orders**.

These customers can be further segmented into **VIP, high-value, regular, and
low-frequency customers** to develop targeted retention and marketing
strategies.

In [84]:
#Task 8 — Customer Segmentation
# Check the customer summary created in Task 7
customer_summary.head()

,Customer ID,Total_Spending,Number_of_Orders,Total_Quantity,Average_Order_Value
5692,18102,580987.04,145,181645,4006.807172
2277,14646,528602.52,151,367193,3500.678940
1789,14156,313437.62,156,164325,2009.215513
2538,14911,291420.81,398,147972,732.213090
5050,17450,244784.25,51,83914,4799.691176


In [85]:
# 1. CREATE CUSTOMER SEGMENTS

# Create a function to classify customers based on their number of orders

def classify_customer(orders):
    # Customer with only 1 order
    if orders == 1:
        return "New"
    
    # Customer with 2 to 4 orders
    elif orders <= 4:
        return "Returning"
    
    # Customer with 5 or more orders
    else:
        return "Loyal"


# Apply the function to Number_of_Orders
customer_summary["Customer_Segment"] = (customer_summary["Number_of_Orders"].apply(classify_customer))

# Display the result
customer_summary.head(10)

,Customer ID,Total_Spending,Number_of_Orders,Total_Quantity,Average_Order_Value,Customer_Segment
5692,18102,580987.04,145,181645,4006.807172,Loyal
2277,14646,528602.52,151,367193,3500.678940,Loyal
1789,14156,313437.62,156,164325,2009.215513,Loyal
2538,14911,291420.81,398,147972,732.213090,Loyal
5050,17450,244784.25,51,83914,4799.691176,Loyal
1331,13694,195640.69,143,188201,1368.116713,Loyal
5109,17511,172132.87,60,117174,2868.881167,Loyal
4061,16446,168472.50,2,80997,84236.250000,Returning
4295,16684,147142.77,55,104810,2675.323091,Loyal
68,12415,144458.37,28,91447,5159.227500,Loyal


In [86]:
# 2. NUMBER OF CUSTOMERS IN EACH SEGMENT

segment_customers = (customer_summary["Customer_Segment"].value_counts().reset_index())

# Rename columns
segment_customers.columns = ["Customer_Segment","Number_of_Customers"]

# Display result
segment_customers

,Customer_Segment,Number_of_Customers
0,Loyal,2161
1,Returning,2094
2,New,1623


In [87]:
# 3. REVENUE BY CUSTOMER SEGMENT

segment_revenue = (customer_summary.groupby("Customer_Segment")["Total_Spending"].sum()
    .reset_index().sort_values("Total_Spending", ascending=False))

# Display result
segment_revenue

,Customer_Segment,Total_Spending
0,Loyal,1.458676e+07
2,Returning,2.227773e+06
1,New,5.602726e+05


In [88]:
# 4. AVERAGE ORDER VALUE BY SEGMENT

segment_aov = (customer_summary.groupby("Customer_Segment")["Average_Order_Value"].mean()
    .reset_index().sort_values("Average_Order_Value", ascending=False))

# Display result
segment_aov

,Customer_Segment,Average_Order_Value
0,Loyal,406.434467
2,Returning,394.229006
1,New,345.207986


In [89]:
# 5. AVERAGE ORDERS PER CUSTOMER

segment_orders = (customer_summary.groupby("Customer_Segment")["Number_of_Orders"].mean()
    .reset_index().sort_values("Number_of_Orders", ascending=False))

# Display result
segment_orders

,Customer_Segment,Number_of_Orders
0,Loyal,13.661268
2,Returning,2.781280
1,New,1.000000


In [90]:
# 6. COMPLETE CUSTOMER SEGMENT SUMMARY

segment_summary = (customer_summary.groupby("Customer_Segment").agg(
        Number_of_Customers=("Customer ID", "nunique"),Total_Revenue=("Total_Spending", "sum"),
        Average_Spending=("Total_Spending", "mean"),Average_Order_Value=("Average_Order_Value", "mean"),
        Average_Orders=("Number_of_Orders", "mean")).reset_index())

# Sort by revenue
segment_summary = segment_summary.sort_values("Total_Revenue",ascending=False)

# Round numerical values
segment_summary = segment_summary.round(2)

# Display result
display(segment_summary)

,Customer_Segment,Number_of_Customers,Total_Revenue,Average_Spending,Average_Order_Value,Average_Orders
0,Loyal,2161,14586758.33,6750.00,406.43,13.66
2,Returning,2094,2227773.37,1063.88,394.23,2.78
1,New,1623,560272.56,345.21,345.21,1.00


In [91]:
# 7. REVENUE CONTRIBUTION %

# Calculate the percentage of total revenue contributed by each customer segment

total_segment_revenue = segment_summary["Total_Revenue"].sum()

segment_summary["Revenue_Contribution_%"] = (segment_summary["Total_Revenue"]
                    / total_segment_revenue) * 100

# Round the percentage
segment_summary["Revenue_Contribution_%"] = (segment_summary["Revenue_Contribution_%"].round(2))

display(segment_summary)

,Customer_Segment,Number_of_Customers,Total_Revenue,Average_Spending,Average_Order_Value,Average_Orders,Revenue_Contribution_%
0,Loyal,2161,14586758.33,6750.00,406.43,13.66,83.95
2,Returning,2094,2227773.37,1063.88,394.23,2.78,12.82
1,New,1623,560272.56,345.21,345.21,1.00,3.22


Customers are segmented based on their number of orders to understand differences in purchasing behavior. Customers with one order are classified as New, customers with 2–4 orders as Returning, and customers with 5 or more orders as Loyal. The segments are compared based on customer count, revenue, average spending, average order value, and average number of orders.

In [92]:
#Task 9 — Customer Purchase Behaviour
# 1. NUMBER OF ORDERS PER CUSTOMER

# Sort customers based on the number of orders from highest to lowest

orders_per_customer = (customer_summary.sort_values("Number_of_Orders", ascending=False))

# Display the top 10 customers
display(orders_per_customer[["Customer ID", "Number_of_Orders"]].head(10))

,Customer ID,Number_of_Orders
2538,14911,398
400,12748,336
5433,17841,211
2935,15311,208
739,13089,203
2237,14606,192
1789,14156,156
5442,17850,155
2277,14646,151
5692,18102,145


In [93]:
# 2. TOTAL SPENDING PER CUSTOMER

# Sort customers by total spending

spending_per_customer = (customer_summary.sort_values("Total_Spending", ascending=False))

# Display top 10 customers
display(spending_per_customer[["Customer ID", "Total_Spending"]].head(10))

,Customer ID,Total_Spending
5692,18102,580987.04
2277,14646,528602.52
1789,14156,313437.62
2538,14911,291420.81
5050,17450,244784.25
1331,13694,195640.69
5109,17511,172132.87
4061,16446,168472.50
4295,16684,147142.77
68,12415,144458.37


In [94]:
# 3. AVERAGE ORDER VALUE PER CUSTOMER

# Sort customers by their average order value

aov_per_customer = (customer_summary.sort_values("Average_Order_Value", ascending=False))

# Display top 10 customers
display(aov_per_customer[["Customer ID", "Average_Order_Value"]].head(10))

,Customer ID,Average_Order_Value
4061,16446,84236.250000
3371,15749,14844.766667
2722,15098,13305.500000
1324,13687,11880.840000
569,12918,10953.500000
5642,18052,10877.180000
3382,15760,6958.170000
1538,13902,6819.052000
0,12346,6463.038333
11,12357,6095.886667


In [95]:
# 4. TOTAL QUANTITY PURCHASED

# Sort customers by total quantity purchased

quantity_per_customer = (customer_summary.sort_values("Total_Quantity", ascending=False))

# Display top 10 customers
display(quantity_per_customer[["Customer ID", "Total_Quantity"]].head(10))

,Customer ID,Total_Quantity
2277,14646,367193
1538,13902,220600
1331,13694,188201
5692,18102,181645
1789,14156,164325
2538,14911,147972
5109,17511,117174
4295,16684,104810
1931,14298,100272
68,12415,91447


In [96]:
# 5. PURCHASE FREQUENCY

# Calculate the average number of orders across all customers

average_orders = customer_summary["Number_of_Orders"].mean()

print(f"Average Orders per Customer: " f"{average_orders:.2f}")

Average Orders per Customer: 6.29


In [97]:
# Create a column to identify high-frequency customers

customer_summary["Purchase_Frequency"] = customer_summary["Number_of_Orders"] > average_orders

# Display customers with above-average frequency
display(customer_summary[customer_summary["Purchase_Frequency"]].head(10))

,Customer ID,Total_Spending,Number_of_Orders,Total_Quantity,Average_Order_Value,Customer_Segment,Purchase_Frequency
5692,18102,580987.04,145,181645,4006.807172,Loyal,True
2277,14646,528602.52,151,367193,3500.678940,Loyal,True
1789,14156,313437.62,156,164325,2009.215513,Loyal,True
2538,14911,291420.81,398,147972,732.213090,Loyal,True
5050,17450,244784.25,51,83914,4799.691176,Loyal,True
1331,13694,195640.69,143,188201,1368.116713,Loyal,True
5109,17511,172132.87,60,117174,2868.881167,Loyal,True
4295,16684,147142.77,55,104810,2675.323091,Loyal,True
68,12415,144458.37,28,91447,5159.227500,Loyal,True
2685,15061,126389.02,127,74280,995.189134,Loyal,True


In [98]:
# 6. ORDERS VS TOTAL SPENDING

# Calculate correlation between: Number of Orders and Total Spending

correlation = customer_summary[["Number_of_Orders", "Total_Spending"]].corr()

display(correlation)

,Number_of_Orders,Total_Spending
Number_of_Orders,1.00000,0.62792
Total_Spending,0.62792,1.00000


In [99]:
# 7. ORDERS VS AVERAGE ORDER VALUE

orders_aov_correlation = customer_summary[["Number_of_Orders", "Average_Order_Value"]].corr()

display(orders_aov_correlation)

,Number_of_Orders,Average_Order_Value
Number_of_Orders,1.000000,0.033768
Average_Order_Value,0.033768,1.000000


In [100]:
# 8. CUSTOMER BEHAVIOUR SUMMARY

customer_behavior = customer_summary[["Customer ID","Number_of_Orders","Total_Spending",
        "Average_Order_Value","Total_Quantity","Customer_Segment"]].copy()

# Display the first 10 customers
display(customer_behavior.head(10))

,Customer ID,Number_of_Orders,Total_Spending,Average_Order_Value,Total_Quantity,Customer_Segment
5692,18102,145,580987.04,4006.807172,181645,Loyal
2277,14646,151,528602.52,3500.678940,367193,Loyal
1789,14156,156,313437.62,2009.215513,164325,Loyal
2538,14911,398,291420.81,732.213090,147972,Loyal
5050,17450,51,244784.25,4799.691176,83914,Loyal
1331,13694,143,195640.69,1368.116713,188201,Loyal
5109,17511,60,172132.87,2868.881167,117174,Loyal
4061,16446,2,168472.50,84236.250000,80997,Returning
4295,16684,55,147142.77,2675.323091,104810,Loyal
68,12415,28,144458.37,5159.227500,91447,Loyal


In [101]:
# 9. HIGHEST-VALUE CUSTOMER

# Find the customer who spent the most

highest_value_customer = customer_summary.loc[customer_summary["Total_Spending"].idxmax()]
display(highest_value_customer)

Customer ID                  18102
Total_Spending           580987.04
Number_of_Orders               145
Total_Quantity              181645
Average_Order_Value    4006.807172
Customer_Segment             Loyal
Purchase_Frequency            True
Name: 5692, dtype: object

In [102]:
# 10. MOST FREQUENT CUSTOMER

# Find the customer with the highest number of orders

most_frequent_customer = customer_summary.loc[customer_summary["Number_of_Orders"].idxmax()]
display(most_frequent_customer)

Customer ID                14911
Total_Spending         291420.81
Number_of_Orders             398
Total_Quantity            147972
Average_Order_Value    732.21309
Customer_Segment           Loyal
Purchase_Frequency          True
Name: 2538, dtype: object

In [103]:
# 11. CUSTOMER WITH HIGHEST AOV


# Find the customer with the highest average order value

highest_aov_customer = customer_summary.loc[customer_summary["Average_Order_Value"].idxmax()]
display(highest_aov_customer)

Customer ID                16446
Total_Spending          168472.5
Number_of_Orders               2
Total_Quantity             80997
Average_Order_Value     84236.25
Customer_Segment       Returning
Purchase_Frequency         False
Name: 4061, dtype: object

Task 9: Customer Purchase Behaviour

Analyze customer purchasing patterns based on order frequency, total spending, average order value, and total quantity purchased. The analysis identifies the most frequent, highest-value, and highest-AOV customers and examines the relationship between purchase frequency and customer spending.

In [104]:
#Task 10: Country-wise Sales Analysis

country_summary = (df.groupby("Country").agg(Total_Revenue=("Total_Sales", "sum"),
          Total_Quantity=("Quantity", "sum"),Number_of_Orders=("Invoice", "nunique"),
          Number_of_Customers=("Customer ID", "nunique")).reset_index())

country_summary.head()

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers
0,Australia,169283.46,103759,95,15
1,Austria,23613.01,11578,45,13
2,Bahrain,1354.37,755,4,2
3,Belgium,65387.82,34768,149,29
4,Brazil,1411.87,545,2,2


In [105]:
#Top 10 Countries by Revenue
top_countries_revenue = (country_summary.sort_values("Total_Revenue", ascending=False)
    .head(10))
top_countries_revenue

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers
38,United Kingdom,1.438923e+07,8532411,33541,5350
10,EIRE,6.165705e+05,318021,567,5
24,Netherlands,5.540381e+05,383879,228,22
14,Germany,4.250197e+05,225154,789,107
13,France,3.487690e+05,270288,614,95
0,Australia,1.692835e+05,103759,95,15
32,Spain,1.083325e+05,50307,154,41
34,Switzerland,1.000619e+05,52227,90,22
33,Sweden,9.151582e+04,88495,104,19
9,Denmark,6.858069e+04,237471,43,12


In [106]:
#Top 10 Countries by Number of Orders
top_countries_orders = (country_summary.sort_values("Number_of_Orders", ascending=False)
    .head(10))
top_countries_orders

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers
38,United Kingdom,1.438923e+07,8532411,33541,5350
14,Germany,4.250197e+05,225154,789,107
13,France,3.487690e+05,270288,614,95
10,EIRE,6.165705e+05,318021,567,5
24,Netherlands,5.540381e+05,383879,228,22
32,Spain,1.083325e+05,50307,154,41
3,Belgium,6.538782e+04,34768,149,29
33,Sweden,9.151582e+04,88495,104,19
0,Australia,1.692835e+05,103759,95,15
28,Portugal,5.555478e+04,27423,93,24


In [107]:
#Top 10 Countries by Number of Customers
top_countries_customers = (country_summary.sort_values("Number_of_Customers", ascending=False)
    .head(10))
top_countries_customers

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers
38,United Kingdom,1.438923e+07,8532411,33541,5350
14,Germany,4.250197e+05,225154,789,107
13,France,3.487690e+05,270288,614,95
32,Spain,1.083325e+05,50307,154,41
3,Belgium,6.538782e+04,34768,149,29
28,Portugal,5.555478e+04,27423,93,24
24,Netherlands,5.540381e+05,383879,228,22
34,Switzerland,1.000619e+05,52227,90,22
33,Sweden,9.151582e+04,88495,104,19
18,Italy,3.210817e+04,15312,65,17


In [108]:
#Calculate Country-wise AOV

#AOV = Total Revenue ÷ Number of Orders

country_summary["Average_Order_Value"] = (country_summary["Total_Revenue"]/ country_summary["Number_of_Orders"])

country_summary = country_summary.round(2)
country_summary.head()

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers,Average_Order_Value
0,Australia,169283.46,103759,95,15,1781.93
1,Austria,23613.01,11578,45,13,524.73
2,Bahrain,1354.37,755,4,2,338.59
3,Belgium,65387.82,34768,149,29,438.84
4,Brazil,1411.87,545,2,2,705.94


In [109]:
#Top 10 Countries by AOV
top_countries_aov = (country_summary.sort_values("Average_Order_Value", ascending=False)
    .head(10))
top_countries_aov

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers,Average_Order_Value
24,Netherlands,554038.09,383879,228,22,2429.99
31,Singapore,25317.06,6994,11,1,2301.55
0,Australia,169283.46,103759,95,15,1781.93
21,Lebanon,1693.88,386,1,1,1693.88
9,Denmark,68580.69,237471,43,12,1594.90
35,Thailand,3070.54,2552,2,1,1535.27
17,Israel,10415.24,5175,7,4,1487.89
19,Japan,43023.91,31643,33,10,1303.75
26,Norway,56322.50,23621,45,13,1251.61
34,Switzerland,100061.94,52227,90,22,1111.80


In [110]:
#Country Revenue Contribution %

total_revenue = country_summary["Total_Revenue"].sum()

country_summary["Revenue_Contribution_%"]=(country_summary["Total_Revenue"] / total_revenue)*100

country_summary = country_summary.round(2)

country_summary.sort_values("Revenue_Contribution_%",ascending=False).head(10)

,Country,Total_Revenue,Total_Quantity,Number_of_Orders,Number_of_Customers,Average_Order_Value,Revenue_Contribution_%
38,United Kingdom,14389234.92,8532411,33541,5350,429.00,82.82
10,EIRE,616570.54,318021,567,5,1087.43,3.55
24,Netherlands,554038.09,383879,228,22,2429.99,3.19
14,Germany,425019.71,225154,789,107,538.68,2.45
13,France,348768.96,270288,614,95,568.03,2.01
0,Australia,169283.46,103759,95,15,1781.93,0.97
32,Spain,108332.49,50307,154,41,703.46,0.62
34,Switzerland,100061.94,52227,90,22,1111.80,0.58
33,Sweden,91515.82,88495,104,19,879.96,0.53
9,Denmark,68580.69,237471,43,12,1594.90,0.39


In [111]:
#Find the Best Country
best_country = country_summary.loc[country_summary["Total_Revenue"].idxmax()]
best_country

Country                   United Kingdom
Total_Revenue                14389234.92
Total_Quantity                   8532411
Number_of_Orders                   33541
Number_of_Customers                 5350
Average_Order_Value                429.0
Revenue_Contribution_%             82.82
Name: 38, dtype: object

### Task 10: Country-wise Sales Analysis

Analyze sales performance across different countries to identify the strongest markets.

The analysis compares countries based on:
- Total revenue
- Total quantity sold
- Number of orders
- Number of customers
- Average order value (AOV)
- Revenue contribution percentage

The objective is to identify the most important markets and understand differences in customer activity and purchasing value across countries.

In [112]:
#Task 11: Cancellation Analysis
#Identify cancelled transactions
# Create a copy for cancellation analysis
cancellation_df = df_original.copy()

# Identify cancelled invoices
cancellation_df["Is_Cancelled"] = (cancellation_df["Invoice"].astype(str).str.startswith("C"))

# Count cancelled and non-cancelled transactions
cancellation_df["Is_Cancelled"].value_counts()

Is_Cancelled
False    1047877
True       19494
Name: count, dtype: int64

In [113]:
# Filter cancelled transactions
cancelled_transactions = cancellation_df[cancellation_df["Is_Cancelled"] == True]

# Number of cancelled invoices
cancelled_orders = cancelled_transactions["Invoice"].nunique()
cancelled_orders

8292

In [114]:
# Total quantity associated with cancelled transactions
cancelled_quantity = cancelled_transactions["Quantity"].abs().sum()
cancelled_quantity

np.int64(490994)

In [115]:
# Total unique invoices
total_orders_original = cancellation_df["Invoice"].nunique()

# Cancellation rate
cancellation_rate = (cancelled_orders / total_orders_original) * 100
round(cancellation_rate, 2)

15.46

In [116]:
# Country-wise cancellation analysis
country_cancellations = (cancelled_transactions.groupby("Country").agg(
        Cancelled_Quantity=("Quantity", lambda x: x.abs().sum()),
        Cancelled_Orders=("Invoice", "nunique")).reset_index())

# Top countries by cancelled orders
country_cancellations = (country_cancellations.sort_values("Cancelled_Orders", ascending=False))
country_cancellations.head(10)

,Country,Cancelled_Quantity,Cancelled_Orders
34,United Kingdom,368281,7315
12,Germany,3422,306
8,EIRE,9223,180
11,France,90336,124
3,Belgium,180,34
29,Spain,5651,34
31,Switzerland,507,30
25,Portugal,351,29
16,Italy,192,27
4,Channel Islands,1009,24


In [117]:
# Product-wise cancellation analysis
product_cancellations = (cancelled_transactions.groupby(["StockCode", "Description"]).agg(
        Cancelled_Quantity=("Quantity", lambda x: x.abs().sum()),
        Cancelled_Orders=("Invoice", "nunique")).reset_index())

# Products with the highest cancelled quantity
top_cancelled_products = ( product_cancellations.sort_values("Cancelled_Quantity", ascending=False)
    .head(10))
top_cancelled_products

,StockCode,Description,Cancelled_Quantity,Cancelled_Orders
2177,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,1
1918,23166,MEDIUM CERAMIC TOP STORAGE JAR,74494,10
2518,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,18750,6
297,21088,SET/6 FRUIT SALAD PAPER CUPS,7140,2
301,21096,SET/6 FRUIT SALAD PAPER PLATES,7008,1
3082,M,Manual,5452,496
29,16047,POP ART PEN CASE & PENS,5184,1
2851,85110,BLACK SILVER FLOWER T-LIGHT HOLDER,5040,2
2223,37340,MULTICOLOUR SPRING FLOWER MUG,4996,4
28,16046,TEATIME PEN CASE & PENS,4632,2


In [118]:
#Create cancellation summary
cancellation_summary = pd.DataFrame({"Metric": ["Total Orders","Cancelled Orders",
       "Cancellation Rate (%)","Cancelled Quantity"],
"Value": [total_orders_original,cancelled_orders,round(cancellation_rate, 2),cancelled_quantity]})

cancellation_summary

,Metric,Value
0,Total Orders,53628.00
1,Cancelled Orders,8292.00
2,Cancellation Rate (%),15.46
3,Cancelled Quantity,490994.00


### Task 11: Cancellation Analysis

Analyze cancelled transactions to understand the level of order cancellations in the dataset.

The analysis focuses on:
- Total cancelled orders
- Total cancelled quantity
- Cancellation rate
- Country-wise cancellations
- Products with the highest cancelled quantities

The objective is to identify cancellation patterns and understand which markets and products are most affected by cancellations.

In [119]:
#Task 12: Repeat Customer Analysis
# Classify customers based on number of orders

customer_behavior["Customer_Type"] = customer_behavior["Number_of_Orders"].apply(
    lambda x: "One-time Customer" if x == 1 else "Repeat Customer")
customer_behavior.head()

,Customer ID,Number_of_Orders,Total_Spending,Average_Order_Value,Total_Quantity,Customer_Segment,Customer_Type
5692,18102,145,580987.04,4006.807172,181645,Loyal,Repeat Customer
2277,14646,151,528602.52,3500.678940,367193,Loyal,Repeat Customer
1789,14156,156,313437.62,2009.215513,164325,Loyal,Repeat Customer
2538,14911,398,291420.81,732.213090,147972,Loyal,Repeat Customer
5050,17450,51,244784.25,4799.691176,83914,Loyal,Repeat Customer


In [120]:
# Count customers in each category

customer_type_counts = (customer_behavior["Customer_Type"].value_counts().reset_index())

customer_type_counts.columns = ["Customer_Type","Number_of_Customers"]
customer_type_counts

,Customer_Type,Number_of_Customers
0,Repeat Customer,4255
1,One-time Customer,1623


In [121]:
# Calculate percentage of customers in each category

total_customers = customer_type_counts["Number_of_Customers"].sum()

customer_type_counts["Customer_Percentage_%"] = (customer_type_counts["Number_of_Customers"]
    / total_customers) * 100

customer_type_counts = customer_type_counts.round(2)
customer_type_counts

,Customer_Type,Number_of_Customers,Customer_Percentage_%
0,Repeat Customer,4255,72.39
1,One-time Customer,1623,27.61


In [122]:
# Revenue generated by each customer type

repeat_customer_revenue = (customer_behavior.groupby("Customer_Type").agg(
        Total_Revenue=("Total_Spending", "sum"),Average_Spending=("Total_Spending", "mean"),
        Average_Order_Value=("Average_Order_Value", "mean"),Average_Orders=("Number_of_Orders", "mean")
    ).reset_index())
repeat_customer_revenue

,Customer_Type,Total_Revenue,Average_Spending,Average_Order_Value,Average_Orders
0,One-time Customer,5.602726e+05,345.207986,345.207986,1.000000
1,Repeat Customer,1.681453e+07,3951.711329,400.427831,8.306933


In [123]:
# Calculate total revenue

total_customer_revenue = repeat_customer_revenue["Total_Revenue"].sum()

# Calculate revenue contribution percentage

repeat_customer_revenue["Revenue_Contribution_%"] = (repeat_customer_revenue["Total_Revenue"]
    / total_customer_revenue) * 100
repeat_customer_revenue = repeat_customer_revenue.round(2)
repeat_customer_revenue

,Customer_Type,Total_Revenue,Average_Spending,Average_Order_Value,Average_Orders,Revenue_Contribution_%
0,One-time Customer,560272.56,345.21,345.21,1.00,3.22
1,Repeat Customer,16814531.71,3951.71,400.43,8.31,96.78


In [124]:
# Filter only repeat customers

repeat_customers = customer_behavior[customer_behavior["Customer_Type"] == "Repeat Customer"]

# Top 10 repeat customers by spending

top_repeat_customers = (repeat_customers.sort_values("Total_Spending", ascending=False).head(10))
top_repeat_customers

,Customer ID,Number_of_Orders,Total_Spending,Average_Order_Value,Total_Quantity,Customer_Segment,Customer_Type
5692,18102,145,580987.04,4006.807172,181645,Loyal,Repeat Customer
2277,14646,151,528602.52,3500.678940,367193,Loyal,Repeat Customer
1789,14156,156,313437.62,2009.215513,164325,Loyal,Repeat Customer
2538,14911,398,291420.81,732.213090,147972,Loyal,Repeat Customer
5050,17450,51,244784.25,4799.691176,83914,Loyal,Repeat Customer
1331,13694,143,195640.69,1368.116713,188201,Loyal,Repeat Customer
5109,17511,60,172132.87,2868.881167,117174,Loyal,Repeat Customer
4061,16446,2,168472.50,84236.250000,80997,Returning,Repeat Customer
4295,16684,55,147142.77,2675.323091,104810,Loyal,Repeat Customer
68,12415,28,144458.37,5159.227500,91447,Loyal,Repeat Customer


In [125]:
# Calculate repeat customer rate

repeat_customer_count = (customer_behavior["Customer_Type"].eq("Repeat Customer").sum())

repeat_customer_rate = (repeat_customer_count / total_customers) * 100
round(repeat_customer_rate, 2)

np.float64(72.39)

In [126]:
# Create final repeat customer summary

repeat_summary = pd.DataFrame({"Metric": ["Total Customers","One-time Customers",
                    "Repeat Customers","Repeat Customer Rate (%)"],
    "Value": [total_customers,(customer_behavior["Customer_Type"] == "One-time Customer").sum(),
        repeat_customer_count,round(repeat_customer_rate, 2)]})
repeat_summary

,Metric,Value
0,Total Customers,5878.00
1,One-time Customers,1623.00
2,Repeat Customers,4255.00
3,Repeat Customer Rate (%),72.39


### Task 12: Repeat Customer Analysis

Analyze customer retention by comparing one-time customers with repeat customers.

The analysis focuses on:
- Number of one-time customers
- Number of repeat customers
- Repeat customer rate
- Revenue contribution of repeat customers
- Average spending and order value
- Top repeat customers by total spending

The objective is to understand customer retention and determine how important repeat customers are to overall business revenue.

In [127]:
#Task 13: Monthly Growth Analysis
# Calculate total revenue for each month

monthly_sales = (df.groupby("Year_Month")["Total_Sales"].sum().reset_index())

monthly_sales.columns = ["Year_Month","Total_Revenue"]
monthly_sales

,Year_Month,Total_Revenue
0,2009-12,683504.010
1,2010-01,555802.672
2,2010-02,504558.956
3,2010-03,696978.471
4,2010-04,591982.002
5,2010-05,597833.380
6,2010-06,636371.130
7,2010-07,589736.170
8,2010-08,602224.600
9,2010-09,829013.951


In [128]:
# Calculate month-over-month revenue growth

monthly_sales["Growth_%"] = (monthly_sales["Total_Revenue"].pct_change()* 100)
monthly_sales = monthly_sales.round(2)
monthly_sales

,Year_Month,Total_Revenue,Growth_%
0,2009-12,683504.01,NaN
1,2010-01,555802.67,-18.68
2,2010-02,504558.96,-9.22
3,2010-03,696978.47,38.14
4,2010-04,591982.00,-15.06
5,2010-05,597833.38,0.99
6,2010-06,636371.13,6.45
7,2010-07,589736.17,-7.33
8,2010-08,602224.60,2.12
9,2010-09,829013.95,37.66


In [129]:
# Find the month with the highest growth

highest_growth_month = monthly_sales.loc[monthly_sales["Growth_%"].idxmax()]
highest_growth_month

Year_Month        2011-09
Total_Revenue    950690.2
Growth_%            47.61
Name: 21, dtype: object

In [130]:
# Find the month with the biggest revenue decline

biggest_decline_month = monthly_sales.loc[monthly_sales["Growth_%"].idxmin()]
biggest_decline_month

Year_Month         2011-12
Total_Revenue    517208.44
Growth_%            -55.27
Name: 24, dtype: object

In [131]:
# Find the month with the highest total revenue

highest_revenue_month = monthly_sales.loc[monthly_sales["Total_Revenue"].idxmax()]
highest_revenue_month

Year_Month          2010-11
Total_Revenue    1166460.02
Growth_%              12.91
Name: 11, dtype: object

In [132]:
# Find the month with the lowest total revenue

lowest_revenue_month = monthly_sales.loc[monthly_sales["Total_Revenue"].idxmin()]
lowest_revenue_month

Year_Month         2011-02
Total_Revenue    446084.92
Growth_%            -21.48
Name: 14, dtype: object

In [134]:
# Classify each month based on revenue growth

monthly_sales["Growth_Status"] = monthly_sales["Growth_%"].apply(lambda x: ("Growth" if x > 0
        else "Decline" if x < 0 else "No Change"))
monthly_sales

,Year_Month,Total_Revenue,Growth_%,Growth_Status
0,2009-12,683504.01,NaN,No Change
1,2010-01,555802.67,-18.68,Decline
2,2010-02,504558.96,-9.22,Decline
3,2010-03,696978.47,38.14,Growth
4,2010-04,591982.00,-15.06,Decline
5,2010-05,597833.38,0.99,Growth
6,2010-06,636371.13,6.45,Growth
7,2010-07,589736.17,-7.33,Decline
8,2010-08,602224.60,2.12,Growth
9,2010-09,829013.95,37.66,Growth


In [135]:
# Create a summary of the monthly growth analysis

growth_summary = pd.DataFrame({"Metric": ["Highest Revenue Month","Lowest Revenue Month",
        "Highest Growth Month","Biggest Decline Month"],
    "Value": [highest_revenue_month["Year_Month"],lowest_revenue_month["Year_Month"],
        highest_growth_month["Year_Month"],biggest_decline_month["Year_Month"]]})
growth_summary

,Metric,Value
0,Highest Revenue Month,2010-11
1,Lowest Revenue Month,2011-02
2,Highest Growth Month,2011-09
3,Biggest Decline Month,2011-12


### Task 13: Monthly Growth Analysis

Analyze monthly revenue performance and identify changes in business growth over time.

The analysis calculates:
- Monthly total revenue
- Month-over-month revenue growth
- Highest revenue month
- Lowest revenue month
- Month with the highest growth
- Month with the biggest revenue decline

The objective is to understand the business growth pattern and identify periods of strong performance and declining sales.

In [136]:
#Task 14: Business Insights & Recommendations
# Create a final KPI summary

business_kpis = pd.DataFrame({"KPI": ["Total Revenue","Total Quantity Sold","Total Orders",
        "Total Customers","Total Products","Average Order Value","Average Product Price",
        "Cancellation Rate","Repeat Customer Rate"],
    "Value": [total_revenue,total_quantity,total_orders,total_customers,total_products,
    average_order_value,average_product_price,cancellation_rate,repeat_customer_rate]})

business_kpis.round(2)

,KPI,Value
0,Total Revenue,17374804.27
1,Total Quantity Sold,10513952.00
2,Total Orders,36969.00
3,Total Customers,5878.00
4,Total Products,4631.00
5,Average Order Value,469.98
6,Average Product Price,3.22
7,Cancellation Rate,15.46
8,Repeat Customer Rate,72.39


In [137]:
# Country generating the highest revenue

best_country = country_summary.loc[country_summary["Total_Revenue"].idxmax()]
best_country

Country                   United Kingdom
Total_Revenue                14389234.92
Total_Quantity                   8532411
Number_of_Orders                   33541
Number_of_Customers                 5350
Average_Order_Value                429.0
Revenue_Contribution_%             82.82
Name: 38, dtype: object

In [138]:
# Product with the highest revenue

best_product = product_summary.loc[product_summary["Total_Revenue"].idxmax()]
best_product

StockCode                              22423
Description         REGENCY CAKESTAND 3 TIER
Total_Revenue                      277656.25
Total_Quantity                         24124
Number_of_Orders                        3317
Average_Price                      12.461649
Name: 1846, dtype: object

In [139]:
# Customer with the highest total spending

highest_value_customer

Customer ID                  18102
Total_Spending           580987.04
Number_of_Orders               145
Total_Quantity              181645
Average_Order_Value    4006.807172
Customer_Segment             Loyal
Purchase_Frequency            True
Name: 5692, dtype: object

In [140]:
# Find revenue contribution from repeat customers

repeat_revenue = repeat_customer_revenue[repeat_customer_revenue["Customer_Type"] == "Repeat Customer"]
repeat_revenue

,Customer_Type,Total_Revenue,Average_Spending,Average_Order_Value,Average_Orders,Revenue_Contribution_%
1,Repeat Customer,16814531.71,3951.71,400.43,8.31,96.78


In [141]:
# Month with the highest revenue growth

highest_growth_month

Year_Month        2011-09
Total_Revenue    950690.2
Growth_%            47.61
Name: 21, dtype: object

In [142]:
# Month with the biggest revenue decline

biggest_decline_month

Year_Month         2011-12
Total_Revenue    517208.44
Growth_%            -55.27
Name: 24, dtype: object

In [143]:
# Business recommendations

recommendations = ["Focus marketing efforts on the highest-revenue countries.",
    "Promote top-performing products to increase revenue further.",
    "Develop loyalty programs to increase repeat purchases.",
    "Investigate reasons behind cancelled orders and reduce cancellation rates.",
    "Study high-value customers and create targeted offers for them.",
    "Analyze high-growth months to identify successful sales strategies.",
    "Investigate periods of declining sales and identify possible causes."]

recommendations_df = pd.DataFrame({"Recommendation": recommendations})

recommendations_df

,Recommendation
0,Focus marketing efforts on the highest-revenue...
1,Promote top-performing products to increase re...
2,Develop loyalty programs to increase repeat pu...
3,Investigate reasons behind cancelled orders an...
4,Study high-value customers and create targeted...
5,Analyze high-growth months to identify success...
6,Investigate periods of declining sales and ide...


### Task 14: Business Insights & Recommendations

#### Key Business Insights

1. **Strong UK Market:**  
   The United Kingdom is the dominant market, generating £14.39 million in revenue and contributing 82.82% of total revenue. This indicates that the business is highly dependent on the UK market.

2. **Top-performing Product:**  
   The REGENCY CAKESTAND 3 TIER is the highest-revenue-generating product, generating £277,656.25 in revenue from 24,124 units sold across 3,317 orders.

3. **High-value Customer:**  
   Customer 18102 is the highest-spending customer, with total spending of £580,987.04 across 145 orders. This customer also has an average order value of approximately £4,006.81, making them a highly valuable customer.

4. **Repeat Customers Drive Revenue:**  
   Repeat customers generated £16.81 million in revenue, contributing 96.78% of total customer revenue. This indicates that repeat customers are extremely important to the business.

5. **Strong Customer Loyalty:**  
   Repeat customers placed an average of 8.31 orders and had an average spending of approximately £3,951.71, indicating strong purchasing activity among returning customers.

6. **Strongest Growth Month:**  
   September 2011 recorded the highest month-over-month revenue growth of 47.61%, with total revenue of approximately £950,690.

7. **Significant Revenue Decline:**  
   December 2011 experienced the largest month-over-month decline of 55.27%, with revenue of approximately £517,208. This period should be investigated to understand the reasons behind the decline.

#### Business Recommendations

1. The business should continue strengthening its UK market while exploring opportunities to reduce dependence on a single market.

2. High-performing products such as the REGENCY CAKESTAND 3 TIER should be prioritized in promotions, inventory planning, and product recommendations.

3. High-value customers should receive personalized offers, loyalty benefits, and targeted marketing campaigns.

4. Since repeat customers contribute a very large share of revenue, customer retention strategies should be a major business priority.

5. The reasons behind the significant December 2011 revenue decline should be investigated to determine whether it was caused by seasonal trends, reduced demand, cancellations, or other factors.

6. The strategies responsible for the strong growth in September 2011 should be analyzed and, where appropriate, replicated in future campaigns.

## Task 15: Final Project Conclusion

### Project Summary

This project analyzed the Online Retail dataset using Python and Pandas to understand the business's sales performance, product performance, customer behaviour, customer retention, country-wise performance, and cancellation patterns.

The analysis involved data cleaning, feature engineering, aggregation, customer segmentation, trend analysis, and business performance evaluation.

### Key Findings

- The United Kingdom is the dominant market, contributing 82.82% of total revenue.
- The REGENCY CAKESTAND 3 TIER was the highest-revenue-generating product, generating £277,656.25 in revenue.
- Customer 18102 was the highest-spending customer, with total spending of £580,987.04.
- Repeat customers generated £16.81 million in revenue and contributed 96.78% of total customer revenue.
- September 2011 recorded the highest month-over-month revenue growth of 47.61%.
- December 2011 experienced the largest month-over-month revenue decline of 55.27%.

### Overall Business Conclusion

The analysis shows that customer retention plays a critical role in the business, as repeat customers contribute the majority of revenue. The United Kingdom is the primary market and represents a significant source of business revenue.

High-value customers and top-performing products also contribute substantially to overall performance. However, the strong dependence on the UK market indicates an opportunity to explore and develop other markets.

The significant variation in monthly revenue, particularly the strong growth in September and decline in December 2011, suggests that seasonal patterns and other business factors should be investigated further.

Overall, the analysis demonstrates how data-driven decision-making can help the business improve customer retention, focus on high-performing products and markets, reduce potential revenue losses, and identify future growth opportunities.

### Future Scope

Further analysis could include:

- Sales visualization and interactive dashboards
- Customer RFM analysis
- Product profitability analysis
- Customer churn prediction
- Sales forecasting
- Advanced customer segmentation
- Country-level market expansion analysis